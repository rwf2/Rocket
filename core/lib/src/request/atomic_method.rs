use crate::http::Method;

pub struct AtomicMethod(ref_swap::RefSwap<'static, Method>);

#[inline(always)]
const fn makeref(method: Method) -> &'static Method {
    match method {
        Method::Get => &Method::Get,
        Method::Put => &Method::Put,
        Method::Post => &Method::Post,
        Method::Delete => &Method::Delete,
        Method::Options => &Method::Options,
        Method::Head => &Method::Head,
        Method::Trace => &Method::Trace,
        Method::Connect => &Method::Connect,
        Method::Patch => &Method::Patch,
        Method::Any => &Method::Any,
    }
}

impl AtomicMethod {
    pub fn new(value: Method) -> Self {
        Self(ref_swap::RefSwap::new(makeref(value)))
    }

    pub fn load(&self) -> Method {
        *self.0.load(std::sync::atomic::Ordering::Acquire)
    }

    pub fn set(&mut self, new: Method) {
        *self = Self::new(new);
    }

    pub fn store(&self, new: Method) {
        self.0.store(makeref(new), std::sync::atomic::Ordering::Release)
    }
}

impl Clone for AtomicMethod {
    fn clone(&self) -> Self {
        let inner = self.0.load(std::sync::atomic::Ordering::Acquire);
        Self(ref_swap::RefSwap::new(inner))
    }
}
