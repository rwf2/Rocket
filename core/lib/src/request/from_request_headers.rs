use crate::{outcome::IntoOutcome, Request};
use super::FromRequest;

use headers::{Header as HHeader, HeaderValue as HHeaderValue};
use rocket_http::Status;

macro_rules! typed_headers_from_request {
    ($($name:ident),*) => ($(
        pub use crate::http::$name;

        #[rocket::async_trait]
        impl<'r> FromRequest<'r> for $name {
            type Error = headers::Error;
            async fn from_request(req: &'r Request<'_>) ->
                    crate::request::Outcome<Self, Self::Error> {
                req.headers().get($name::name().as_str()).next().or_forward(Status::NotFound)
                    .and_then(|h| HHeaderValue::from_str(h).or_error(Status::BadRequest))
                    .map_error(|(s, _)| (s, headers::Error::invalid()))
                    .and_then(|h| $name::decode(&mut std::iter::once(&h))
                        .or_forward(Status::BadRequest))
            }
        }
    )*)
}

macro_rules! generic_typed_headers_from_request {
($($name:ident<$bound:ident>),*) => ($(
    pub use crate::http::$name;

    #[rocket::async_trait]
    impl<'r, T1: 'static + $bound> FromRequest<'r> for $name<T1> {
        type Error = headers::Error;
        async fn from_request(req: &'r Request<'_>) -> crate::request::Outcome<Self, Self::Error> {
            req.headers().get($name::<T1>::name().as_str()).next()
                .or_forward(Status::NotFound)
                .and_then(|h| HHeaderValue::from_str(h).or_error(Status::BadRequest))
                .map_error(|(s, _)| (s, headers::Error::invalid()))
                .and_then(|h| $name::decode(&mut std::iter::once(&h))
                    .or_forward(Status::BadRequest))
        }
    }
)*)
}

// The following headers from 'headers' 0.4 are not imported, since they are
// provided by other Rocket features.

// * ContentType, // Content-Type header, defined in RFC7231
// * Cookie, // Cookie header, defined in RFC6265
// * Host, // The Host header.
// * Location, // Location header, defined in RFC7231
// * SetCookie, // Set-Cookie header, defined RFC6265

typed_headers_from_request! {
  AcceptRanges, // Accept-Ranges header, defined in RFC7233
  AccessControlAllowCredentials, // Access-Control-Allow-Credentials header, part of CORS
  AccessControlAllowHeaders, // Access-Control-Allow-Headers header, part of CORS
  AccessControlAllowMethods, // Access-Control-Allow-Methods header, part of CORS
  AccessControlAllowOrigin, // The Access-Control-Allow-Origin response header, part of CORS
  AccessControlExposeHeaders, // Access-Control-Expose-Headers header, part of CORS
  AccessControlMaxAge, // Access-Control-Max-Age header, part of CORS
  AccessControlRequestHeaders, // Access-Control-Request-Headers header, part of CORS
  AccessControlRequestMethod, // Access-Control-Request-Method header, part of CORS
  Age, // Age header, defined in RFC7234
  Allow, // Allow header, defined in RFC7231
  CacheControl, // Cache-Control header, defined in RFC7234 with extensions in RFC8246
  Connection, // Connection header, defined in RFC7230
  ContentDisposition, // A Content-Disposition header, (re)defined in RFC6266.
  ContentEncoding, // Content-Encoding header, defined in RFC7231
  ContentLength, // Content-Length header, defined in RFC7230
  ContentLocation, // Content-Location header, defined in RFC7231
  ContentRange, // Content-Range, described in RFC7233
  Date, // Date header, defined in RFC7231
  ETag, // ETag header, defined in RFC7232
  Expect, // The Expect header.
  Expires, // Expires header, defined in RFC7234
  IfMatch, // If-Match header, defined in RFC7232
  IfModifiedSince, // If-Modified-Since header, defined in RFC7232
  IfNoneMatch, // If-None-Match header, defined in RFC7232
  IfRange, // If-Range header, defined in RFC7233
  IfUnmodifiedSince, // If-Unmodified-Since header, defined in RFC7232
  LastModified, // Last-Modified header, defined in RFC7232
  Origin, // The Origin header.
  Pragma, // The Pragma header defined by HTTP/1.0.
  Range, // Range header, defined in RFC7233
  Referer, // Referer header, defined in RFC7231
  ReferrerPolicy, // Referrer-Policy header, part of Referrer Policy
  RetryAfter, // The Retry-After header.
  SecWebsocketAccept, // The Sec-Websocket-Accept header.
  SecWebsocketKey, // The Sec-Websocket-Key header.
  SecWebsocketVersion, // The Sec-Websocket-Version header.
  Server, // Server header, defined in RFC7231
  StrictTransportSecurity, // StrictTransportSecurity header, defined in RFC6797
  Te, // TE header, defined in RFC7230
  TransferEncoding, // Transfer-Encoding header, defined in RFC7230
  Upgrade, // Upgrade header, defined in RFC7230
  UserAgent, // User-Agent header, defined in RFC7231
  Vary // Vary header, defined in RFC7231
}

pub use headers::authorization::Credentials;

generic_typed_headers_from_request! {
    Authorization<Credentials>, // Authorization header, defined in RFC7235
    ProxyAuthorization<Credentials> // Proxy-Authorization header, defined in RFC7235
}
