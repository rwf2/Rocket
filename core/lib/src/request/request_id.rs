use std::{
    fmt::{Debug, Display},
    num::NonZeroU64,
    sync::atomic::AtomicU64, convert::Infallible,
};

use super::{FromRequest, Outcome, Request};

/// Opaque Request ID type. Every incoming request is assigned a unique ID, which can be used to
/// identify which log messages are related to a specific request
///
/// NOTE:
///     Although the RequestId itself is unique for every request, the `Display` implementation
///     only prints part of the whole id. This is enough to differentiate requests nearby, but not
///     enough to differntiate all requests Rocket serves.
///
/// This type is designed only for use when logging. As such, there are no methods provided to
/// get the internal id value, just `Display` & `Debug` implementations. Future implementations may
/// choose to implement more functionality, such as `Hash` & `Eq`.
#[derive(Clone, Copy)]
pub struct RequestId {
    // NonZeroU64, so that Option<RequestId> can be optimized to a single 64 bit value
    id: NonZeroU64,
}

impl RequestId {
    /// Convert a number to an id
    ///
    /// # Panics
    /// panics if the id is zero
    #[cfg(test)]
    pub(crate) fn from(id: u64) -> Self {
        Self { id: NonZeroU64::new(id).unwrap() }
    }
}

const MAX_PRINT_ID: u64 = 0x10000;

impl Display for RequestId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // Avoid printing more than 4 digits by specifying a maximum id. Mod is used to apply a
        // wrapping, i.e. the log will wrap requests periodically. It's unlikely a server can
        // handle 65536 requests concurrently, so this should be enough for to differntiate any
        // request.
        write!(f, "Req {:X}>", self.id.get() % MAX_PRINT_ID)
    }
}

impl Debug for RequestId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Request ID {{ {:X} }}", self.id)
    }
}

#[crate::async_trait]
impl<'r> FromRequest<'r> for RequestId {
    type Error = Infallible;

    async fn from_request(request: & 'r Request<'_>) -> Outcome<Self,Self::Error> {
        Outcome::Success(request.request_id())
    }
}

/// Struct to generate request ids
#[derive(Debug)]
pub(crate) struct RequestIdGenerator {
    next: AtomicU64,
}

impl RequestIdGenerator {
    /// Construct a default generator starting at one
    pub(crate) fn new() -> Self {
        Self {
            next: AtomicU64::new(1),
        }
    }

    /// Get the next ID from the generator
    pub(crate) fn next(&self) -> RequestId {
        // SAFETY: self.next is initalized to one, and only ever incremented. Since self.next is a
        // u64, overflow can be considered impossible.
        RequestId {
            id: unsafe {
                NonZeroU64::new_unchecked(self.next.fetch_add(1, atomic::Ordering::AcqRel))
            },
        }
    }
}

impl Default for RequestIdGenerator {
    fn default() -> Self {
        Self::new()
    }
}

/// Get the id of the current request, if such a request exists.
///
/// ```rust
/// #[get("/")]
/// fn get_id() -> String {
///     format!("{}", current_request().unwrap())
/// }
/// ```
// TODO: add test to doc comment
pub fn current_request() -> Option<RequestId> {
    CURRENT_REQUEST.try_with(|id| *id).ok()
}

tokio::task_local! {
    pub static CURRENT_REQUEST: RequestId;
}
