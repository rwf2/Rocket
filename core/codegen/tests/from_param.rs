use rocket::request::FromParam;

#[allow(non_camel_case_types)]
#[derive(Debug, FromParam, PartialEq)]
enum Test {
    Test1,
    Test2,
    r#for,
}

#[test]
fn derive_from_param() {
    assert_eq!(Test::from_param("Test1").unwrap(), Test::Test1);
    assert_eq!(Test::from_param("Test2").unwrap(), Test::Test2);
    assert_eq!(Test::from_param("for").unwrap(), Test::r#for);

    let err = Test::from_param("For").unwrap_err();
    assert_eq!(err.value, "For");
    assert_eq!(err.options, &["Test1", "Test2", "for"]);

    let err = Test::from_param("not_test").unwrap_err();
    assert_eq!(err.value, "not_test");
    assert_eq!(err.options, &["Test1", "Test2", "for"]);

}
