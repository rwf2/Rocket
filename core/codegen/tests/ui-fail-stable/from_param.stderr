error: named structs are not supported
 --> tests/ui-fail-stable/from_param.rs:4:1
  |
4 | / struct Foo1 {
5 | |     a: String
6 | | }
  | |_^

error: [note] error occurred while deriving `FromParam`
 --> tests/ui-fail-stable/from_param.rs:3:10
  |
3 | #[derive(FromParam)]
  |          ^^^^^^^^^
  |
  = note: this error originates in the derive macro `FromParam` (in Nightly builds, run with -Z macro-backtrace for more info)

error: named structs are not supported
 --> tests/ui-fail-stable/from_param.rs:9:1
  |
9 | struct Foo2 {}
  | ^^^^^^^^^^^^^^

error: [note] error occurred while deriving `FromParam`
 --> tests/ui-fail-stable/from_param.rs:8:10
  |
8 | #[derive(FromParam)]
  |          ^^^^^^^^^
  |
  = note: this error originates in the derive macro `FromParam` (in Nightly builds, run with -Z macro-backtrace for more info)

error: variants with data fields are not supported
  --> tests/ui-fail-stable/from_param.rs:13:6
   |
13 |     A(String),
   |      ^^^^^^^^

error: [note] error occurred while deriving `FromParam`
  --> tests/ui-fail-stable/from_param.rs:11:10
   |
11 | #[derive(FromParam)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `FromParam` (in Nightly builds, run with -Z macro-backtrace for more info)

error: tuple structs are not supported
  --> tests/ui-fail-stable/from_param.rs:18:1
   |
18 | struct Foo4(usize);
   | ^^^^^^^^^^^^^^^^^^^

error: [note] error occurred while deriving `FromParam`
  --> tests/ui-fail-stable/from_param.rs:17:10
   |
17 | #[derive(FromParam)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `FromParam` (in Nightly builds, run with -Z macro-backtrace for more info)
